test/gst-vaapi/encode/12bit/hevc.py

```diff
@@ -0,0 +1,99 @@
+###
+### Copyright (C) 2023 Intel Corporation
+###
+### SPDX-License-Identifier: BSD-3-Clause
+###
+
+from .....lib import *
+from .....lib.gstreamer.vaapi.util import *
+from .....lib.gstreamer.vaapi.encoder import EncoderTest
+
+spec = load_test_spec("hevc", "encode", "12bit")
+
+@slash.requires(*have_gst_element("vaapih265dec"))
+@slash.requires(*have_gst_element("vaapih265enc"))
+class HEVC12EncoderBaseTest(EncoderTest):
+  def before(self):
+    super().before()
+    vars(self).update(
+      codec         = "hevc-12",
+      gstdecoder    = "vaapih265dec",
+      gstencoder    = "vaapih265enc",
+      gstmediatype  = "video/x-h265",
+      gstparser     = "h265parse",
+    )
+
+  def get_file_ext(self):
+    return "h265"
+
+@slash.requires(*platform.have_caps("encode", "hevc_12"))
+class HEVC12EncoderTest(HEVC12EncoderBaseTest):
+  def before(self):
+    super().before()
+    vars(self).update(
+      caps      = platform.get_caps("encode", "hevc_12"),
+      lowpower  = False,
+    )
+
+class cqp(HEVC12EncoderTest):
+  def init(self, tspec, case, gop, slices, bframes, qp, quality, profile):
+    vars(self).update(tspec[case].copy())
+    vars(self).update(
+      bframes   = bframes,
+      case      = case,
+      gop       = gop,
+      profile   = profile,
+      qp        = qp,
+      quality   = quality,
+      rcmode    = "cqp",
+      slices    = slices,
+    )
+
+  @slash.parametrize(*gen_hevc_cqp_parameters(spec, ['main12']))
+  def test(self, case, gop, slices, bframes, qp, quality, profile):
+    self.init(spec, case, gop, slices, bframes, qp, quality, profile)
+    self.encode()
+
+class cbr(HEVC12EncoderTest):
+  def init(self, tspec, case, gop, slices, bframes, bitrate, fps, profile):
+    vars(self).update(tspec[case].copy())
+    vars(self).update(
+      bframes   = bframes,
+      bitrate   = bitrate,
+      case      = case,
+      fps       = fps,
+      gop       = gop,
+      maxrate   = bitrate,
+      minrate   = bitrate,
+      profile   = profile,
+      rcmode    = "cbr",
+      slices    = slices,
+    )
+
+  @slash.parametrize(*gen_hevc_cbr_parameters(spec, ['main12']))
+  def test(self, case, gop, slices, bframes, bitrate, fps, profile):
+    self.init(spec, case, gop, slices, bframes, bitrate, fps, profile)
+    self.encode()
+
+class vbr(HEVC12EncoderTest):
+  def init(self, tspec, case, gop, slices, bframes, bitrate, fps, quality, refs, profile):
+    vars(self).update(tspec[case].copy())
+    vars(self).update(
+      bframes   = bframes,
+      bitrate   = bitrate,
+      case      = case,
+      fps       = fps,
+      gop       = gop,
+      maxrate   = int(bitrate / 0.7),
+      minrate   = bitrate,
```

Comment:
If so, minrate will be greater than maxrate, right ? 

I guess this was copied & pasted from 10bit, but according to the comment in commit 4767de24faa96ce26319ed8c4e0a64dd7d593bd4,  Both maxrate and minrate were set incorrectly for 10bit.
+      ## gst-vaapi sets max-bitrate = bitrate and min-bitrate = bitrate * 0.70
+      maxrate      = int(bitrate / 0.7),
+      minrate      = bitrate,

Reply:
Yes, I understand how this could alarm you or be a little confusing ;)

From the vaapi-fits perspective, we want "bitrate" to be normalized from the user test case config and when internally validating the bitrate gap.  That is, when user defines a vaapi-fits test case with a specific bitrate, the goal is to have the libva driver receive the same values (bits_per_second and target_percentage), regardless of which plugins and middleware is handling it.

Unfortunately, each plugin/middleware has different properties and/or applies different rules/functions to the intermediate values before sending to the libva driver.  Originally, gst-vaapi hard-coded the target-percentage to 70%... but later we added this property to gst-vaapi so we could configure it.  However, we kept it at 70% in the test framework to avoid needing to re-baseline all of our PSNR refs.  Either way, here is how the values are propagated from vaapi-fits to the gst-vaapi/gst-va plugins to achieve bitrate normalization (gst-msdk omitted here as an exercise to reader, if interested).

For gst-vaapi:
```
 test.bitrate = user.bitrate
 test.maxrate = test.bitrate / 0.7
 test.minrate = test.bitrate
 gst-vaapi.target_percentage = test.minrate / test.maxrate * 100
 gst-vaapi.bitrate = test.maxrate

 # inside gstreamer vaapi encoder
 libva.bits_per_second = gst-vaapi.bitrate * 1000
 libva.target_percentage = gst-vaapi.target_percentage
```

For gst-va:
```
 test.bitrate = user.bitrate
 test.maxrate = test.bitrate / 0.7
 test.minrate = test.bitrate
 gst-va.target_percentage = test.minrate / test.maxrate * 100
 gst-va.bitrate = test.minrate

 # inside gstreamer va encoder
 libva.bits_per_second = (gst-va.bitrate * 100 / gst-va.target_percentage) * 1000
 libva.target_percentage = gst-va.target_percentage
```

So, for example, if user sets `bitrate = 500` (user.bitrate) at test config, then `libva.target_percentage = 70` and `libva.bits_per_second = 714` for both gst-vaapi and gst-va plugins.  And similarly, should be for gst-msdk too.

Reply:
Short answer:
No.  The `maxrate` will be greater than `minrate`.  Note the division by a fraction makes the `maxrate` greater.  For example, bitrate=500, **maxrate** = 500 / 0.7 = **714**, **minrate** = bitrate = **500**.

Reply:
I am sorry I made a stupid error. I took it as 'maxrate=int(bitrate * 0.7).